### ISSUE TITLE
catalog-import plugin does not support Unicode characters in catalog-info.yaml file

### ISSUE THREAD

**Issue by tylermarien** [backstage/backstage#4276]
## Expected Behavior
Importing a GitHub repo should allow Unicode characters in the submitted catalog-info.yaml file. For example, a file like this should work:
```YAML
apiVersion: backstage.io/v1alpha1
kind: Component
metadata:
  name: test-service
  description: Description with emojis 🤖
spec:
  lifecycle: production
  type: service
  owner: test-team
```

## Current Behavior
Right now if you try to import the file above, you will get the following error: `Failed to submit PR to repo, Failed to execute 'btoa' on 'Window': The string to be encoded contains characters outside of the Latin1 range.`

## Possible Solution
Use a library like [js-base64](https://www.npmjs.com/package/js-base64) to encode the file contents, which allows for Unicode characters.

## Steps to Reproduce
1. Create a custom LocationAnalyzer
2. Modify the generated Entity in the custom LocationAnalyzer by adding a description field with Unicode characters in it
3. Use custom LocationAnalyzer
4. Try to import GitHub repository via `/catalog-import`

## Context
We are trying to add some extra information to the generated catalog-info.yaml file. This reduces the work for developers as they can import the file and not make any modifications afterwards. We fetch the description from GitHub and add it to the generated catalog-info.yaml file. Some of our repos have emojis in them, so creating the PR fails due to the lack of Unicode support in the native `btoa` function.

**Comment by freben** [comment]
Interesting, this would actually be in the import plugin. @dtuite maybe this is something your team might want to have a look at?

**Comment by dtuite** [comment]
@padraigobrien Have you got capacity to take this up?

**Comment by padraigobrien** [comment]
Yes, feel free to assign to me

**Comment by freben** [comment]
Thank you! 👍

**Comment by Fox32** [comment]
Maybe we should wait #4271 till is completed, because that might result in conflicts

**Comment by padraigobrien** [comment]
I am happy to wait, I can be working on the issue quietly in the background and implement once #4271 is merged

**Comment by padraigobrien** [comment]
Hi @tylermarien would it be possible to get more details on how you created the customer Location Analyser, I ask as i have not been able to recreate the problem. I have been able to import the yaml you included.

![image](https://user-images.githubusercontent.com/1910868/106305482-a7632680-6254-11eb-8d19-f199c6d8dbaa.png)

The emoji is displayed, I am assuming that this was the issue?

**Comment by padraigobrien** [comment]
unfortunately i have not been able to reproduce this , I intend to close this issue

**Comment by tylermarien** [comment]
Sorry, I missed the notification. I have created an example of how we modified the LocationAnalyzer. You can view it here: https://github.com/tylermarien/backstage/tree/customized-location-analyzer. Using the code in that branch will throw an error when trying to import a repo like https://github.com/tylermarien/test-repo. It fails when trying to create the PR in GitHub. You can see the error below.

I have also put together a possible fix: https://github.com/tylermarien/backstage/tree/use-js-base64. This uses the js-base64 library to encode file contents before sending them to GitHub (when trying to open the PR). The built-in btoa function does not handle unicode characters.

![Screen Shot 2021-02-04 at 3 55 59 PM](https://user-images.githubusercontent.com/897223/106962277-7ecfa680-6704-11eb-895f-2533d4851f09.png)

**Comment by adamdmharvey** [comment]
@tylermarien Sorry, as a bit of an outsider, pardon the question.  Based on your comment above, do I read correctly that you wrote a custom location analyzer that takes code out of the sample README.md you have in your test repo, and your location analyzer dynamically inserts the record into the catalog.  And it is THIS process you've customized which fails?  And then the fix you provide with the Node library is what helps address it?

If I have that correct, it would seem this is a code fix IN your location analyzer, as opposed to in Backstage's core ? or do i Have that wrong?  Almost wonder if what you found might be useful for addition to the `contrib/` folder, i.e. info about how you wrote your custom analyzer and what it does, but including your unicode supporting code as well?

Thoughts?

**Comment by andrewthauer** [comment]
@adamdmharvey - @tylermarien and I work together so I'll do my best to explain :) 

The `catalog-import` makes a backend call to the `/api/catalog/analyze-location` route. By default the `catalog-backend` is setup to use the `RepoLocationAnalyzer` as a template which the `catalog-import` then uses. However, you can actually swap out the implementation (in `packages/backend/src/plugins/catalog.ts`) which is what we've done. Ours makes a GitHub API call to get the repo metdata. It pre-populates the `metadata.description` based on the repo description. So basically the `/analyze-location` route just passes back the description with an emojii to the `catalog-import` front-end plugin. The plugin actually makes the requests to the GH API. So it seems like an issue on the front-end.

I suspect, if the `description` was editable, you could probably get the same result by adding an emojii or other unicide characters. Hope that helps clarify things.

**Comment by adamdmharvey** [comment]
Great, thanks for the clarifications! Given this, @padraigobrien is this something you're able/willing to still work on given the clarifications, or should we open it to others to contribute?

**Comment by andrewthauer** [comment]
I will note, that this isn't a big blocker. We don't have too many repos that fall under this case. You can technically work around this by manually creating the `catalog-info.yaml` file and then importing it. However, it seems reasonable that this should be possible either way.

**Comment by tylermarien** [comment]
Thanks for clarifying @andrewthauer. @adamdmharvey I pinpointed the issue and created a proposed fix here: https://github.com/backstage/backstage/compare/master...tylermarien:use-js-base64. It is a fairly simple fix, but I am not sure if you want to include another library.

**Comment by padraigobrien** [comment]
Hey @adamdmharvey, I will do it now, and thanks for the reminder. I missed the merge to enable me to do this change.

**Comment by padraigobrien** [comment]
Hi @tylermarien, to update, I am trying to recreate the issue and failing in my attempts. With the latest copy of backstage i have been able to import. I am attaching a screenshot with smiley faces :-) 
![image](https://user-images.githubusercontent.com/1910868/111697191-2a543680-882d-11eb-9c9b-60b6e6ec4a21.png)

I imported with and without an existing catalog yaml file.
I also experimented with other emojis.

**Comment by tylermarien** [comment]
Hi @padraigobrien. Sorry for the confusion as I realize the title and description in the first message are not accurate. The problem does not exist with the **Import** process, it exists with the **Open a PR** process. Now that there is an editable form in the **Create Pull Request** step, it is fairly easy to replicate. You can see the steps below to replicate.

**Steps to replicate**
1. Go to **Register an existing component** page (/catalog-import)
2. Enter a **Repository URL** of a GitHub repository that does NOT already have a catalog-info.yaml file (eg. https://github.com/tylermarien/test-repo)
3. Click **Analyze**
4. Under the **Create Pull Request** form that appears, enter "Team 🤖" into the **Entity Owner** field
5. Click **Create PR**

**Expected behavior**
A PR should be opened in the target GitHub repository with the generated catalog-info.yaml file.

**Actual behavior**
An error saying "Failed to execute 'btoa' on 'Window': The string to be encoded contains characters outside of the Latin1 range." is displayed. There is a branch named "backstage-integration" created in the target repository, but no commits and no pull request.

**Proposed solution**
https://github.com/backstage/backstage/compare/master...tylermarien:use-js-base64

**Screenshots**
![Screen Shot 2021-03-18 at 4 57 56 PM](https://user-images.githubusercontent.com/897223/111708471-873ef500-880b-11eb-937f-d9712cf2232d.png)
![Screen Shot 2021-03-18 at 4 58 20 PM](https://user-images.githubusercontent.com/897223/111708475-8908b880-880b-11eb-91a7-fb887fa52a02.png)

**Comment by padraigobrien** [comment]
Thanks, Tyler. I can reproduce this now. I will start working on it

**Comment by benjdlambert** [comment]
Fixed by #5105

**Comment by padraigobrien** [comment]
Fixed by [#5146](https://github.com/backstage/backstage/pull/5146)

**Comment by Fox32** [comment]
Should be fixed with #5146, thanks @padraigobrien.
